### ISSUE TITLE
[2.4.0-beta2] Wrong JSON-LD and JSON single item response - ignores serialization groups?

### ISSUE THREAD

**Issue by JarJak** [api-platform/api-platform#1018]
On 2.4.0-beta1 when I ask for single resource (`/api/contents/1.jsonld`) I get:
```json
{
   "@context":"/api/contexts/Content",
   "@id":"/api/contents/1",
   "@type":"Content",
   "id":1,
   "contentType":"homepage",
   "modifiedAt":"2018-06-19T16:18:20+00:00",
   "publishedAt":"2019-01-06T19:14:53+00:00",
   "fieldValues":{
      "title":"Labore reprehenderit dolorem repellendus asperiores.",
      "image":{
         "filename":"kitten.jpg",
         "alt":"A cute kitten",
         "path":"/thumbs/kitten.jpg?w=240&h=160&area=files"
      },
      "teaser":"Ut qui dolor debitis esse. Maiores alias repellat maxime ut voluptatem. Eum esse reprehenderit expedita necessitatibus. Animi hic eveniet sed et.\n\nEst pariatur incidunt aut iste. Rem aut nihil et ut natus perspiciatis. Dolore veritatis ab sed repellendus quo deleniti qui.\n\nQui laudantium corrupti labore quidem veniam nemo qui. Sit dolore atque et cumque libero. Omnis earum repellendus hic aut vero aut voluptatem. Ea non magni eligendi occaecati numquam.",
      "content":"Minus sed repellendus corporis nemo. Aut aut veniam at aut aliquid. Architecto tempora quia neque numquam voluptas sint est delectus.\n\nUnde voluptatem animi non ut aut dicta. Omnis vero dolorum aliquid laudantium magni asperiores. Et tempora eveniet soluta modi occaecati.\n\nEa dolorum tenetur voluptatum temporibus illo fuga. Quibusdam et doloribus debitis omnis sed. Tempora in aperiam ullam non odit. Praesentium sunt accusantium dolorem commodi labore eum nostrum quia.",
      "contentlink":"Aut ea iusto est iste est odit.",
      "slug":"labore-reprehenderit-dolorem-repellendus-asperiores"
   },
   "taxonomyValues":[

   ],
   "authorName":"Jane Doe",
   "extras":{
      "title":"Labore reprehenderit dolorem repellendus asperiores.",
      "image":{
         "filename":"kitten.jpg",
         "alt":"A cute kitten",
         "path":"/thumbs/kitten.jpg?w=240&h=160&area=files"
      },
      "excerpt":"Labore reprehenderit dolorem repellendus asperiores. Ut qui dolor debitis esse. Maiores alias …",
      "link":"/homepage/labore-reprehenderit-dolorem-repellendus-asperiores",
      "editLink":"/bolt/edit/1"
   }
}
```
Which is correct.

But after bumping version to 2.4.0-beta2 I get:
```json
{
   "@context":"/api/contexts/Content",
   "@id":"/api/contents",
   "@type":"hydra:Collection",
   "hydra:member":[
      1,
      "homepage",
      "labore-reprehenderit-dolorem-repellendus-asperiores",
      {
         "id":3,
         "displayName":"Jane Doe",
         "username":"jane_admin",
         "email":"jane_admin@example.org"
      },
      {
         "title":"Labore reprehenderit dolorem repellendus asperiores.",
         "image":{
            "filename":"kitten.jpg",
            "alt":"A cute kitten",
            "path":"/thumbs/kitten.jpg?w=240&h=160&area=files"
         },
         "teaser":"Ut qui dolor debitis esse. Maiores alias repellat maxime ut voluptatem. Eum esse reprehenderit expedita necessitatibus. Animi hic eveniet sed et.\n\nEst pariatur incidunt aut iste. Rem aut nihil et ut natus perspiciatis. Dolore veritatis ab sed repellendus quo deleniti qui.\n\nQui laudantium corrupti labore quidem veniam nemo qui. Sit dolore atque et cumque libero. Omnis earum repellendus hic aut vero aut voluptatem. Ea non magni eligendi occaecati numquam.",
         "content":"Minus sed repellendus corporis nemo. Aut aut veniam at aut aliquid. Architecto tempora quia neque numquam voluptas sint est delectus.\n\nUnde voluptatem animi non ut aut dicta. Omnis vero dolorum aliquid laudantium magni asperiores. Et tempora eveniet soluta modi occaecati.\n\nEa dolorum tenetur voluptatum temporibus illo fuga. Quibusdam et doloribus debitis omnis sed. Tempora in aperiam ullam non odit. Praesentium sunt accusantium dolorem commodi labore eum nostrum quia.",
         "contentlink":"Aut ea iusto est iste est odit.",
         "slug":"labore-reprehenderit-dolorem-repellendus-asperiores"
      },
      [

      ],
      {
         "title":"Labore reprehenderit dolorem repellendus asperiores.",
         "image":{
            "filename":"kitten.jpg",
            "alt":"A cute kitten",
            "path":"/thumbs/kitten.jpg?w=240&h=160&area=files"
         },
         "excerpt":"Labore reprehenderit dolorem repellendus asperiores. Ut qui dolor debitis esse. Maiores alias …",
         "link":"/homepage/labore-reprehenderit-dolorem-repellendus-asperiores",
         "editLink":"/bolt/edit/1"
      },
      "published",
      "fa-home",
      "2018-03-25T03:58:54+00:00",
      "2018-06-19T16:18:20+00:00",
      "2019-01-06T19:14:53+00:00",
      "2019-02-13T22:01:37+00:00"
   ],
   "hydra:totalItems":13,
   "hydra:search":{
      "@type":"hydra:IriTemplate",
      "hydra:template":"/api/contents/1.jsonld{?id,id[],contentType,contentType[],status,status[],createdAt,createdAt[],modifiedAt,modifiedAt[],publishedAt,publishedAt[],depublishedAt,depublishedAt[]}",
      "hydra:variableRepresentation":"BasicRepresentation",
      "hydra:mapping":[
         {
            "@type":"IriTemplateMapping",
            "variable":"id",
            "property":"id",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"id[]",
            "property":"id",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"contentType",
            "property":"contentType",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"contentType[]",
            "property":"contentType",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"status",
            "property":"status",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"status[]",
            "property":"status",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"createdAt",
            "property":"createdAt",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"createdAt[]",
            "property":"createdAt",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"modifiedAt",
            "property":"modifiedAt",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"modifiedAt[]",
            "property":"modifiedAt",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"publishedAt",
            "property":"publishedAt",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"publishedAt[]",
            "property":"publishedAt",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"depublishedAt",
            "property":"depublishedAt",
            "required":false
         },
         {
            "@type":"IriTemplateMapping",
            "variable":"depublishedAt[]",
            "property":"depublishedAt",
            "required":false
         }
      ]
   }
}
```

-----

Also JSON response is wrong in beta2 - looks like it ignores serialization groups.

Response in beta2 (invalid):
```json
{
   "id":1,
   "contentType":"homepage",
   "slug":"labore-reprehenderit-dolorem-repellendus-asperiores",
   "author":{
      "id":3,
      "displayName":"Jane Doe",
      "username":"jane_admin",
      "email":"jane_admin@example.org"
   },
   "fields":{
      "title":"Labore reprehenderit dolorem repellendus asperiores.",
      "image":{
         "filename":"kitten.jpg",
         "alt":"A cute kitten",
         "path":"/thumbs/kitten.jpg?w=240&h=160&area=files"
      },
      "teaser":"Ut qui dolor debitis esse. Maiores alias repellat maxime ut voluptatem. Eum esse reprehenderit expedita necessitatibus. Animi hic eveniet sed et.\n\nEst pariatur incidunt aut iste. Rem aut nihil et ut natus perspiciatis. Dolore veritatis ab sed repellendus quo deleniti qui.\n\nQui laudantium corrupti labore quidem veniam nemo qui. Sit dolore atque et cumque libero. Omnis earum repellendus hic aut vero aut voluptatem. Ea non magni eligendi occaecati numquam.",
      "content":"Minus sed repellendus corporis nemo. Aut aut veniam at aut aliquid. Architecto tempora quia neque numquam voluptas sint est delectus.\n\nUnde voluptatem animi non ut aut dicta. Omnis vero dolorum aliquid laudantium magni asperiores. Et tempora eveniet soluta modi occaecati.\n\nEa dolorum tenetur voluptatum temporibus illo fuga. Quibusdam et doloribus debitis omnis sed. Tempora in aperiam ullam non odit. Praesentium sunt accusantium dolorem commodi labore eum nostrum quia.",
      "contentlink":"Aut ea iusto est iste est odit.",
      "slug":"labore-reprehenderit-dolorem-repellendus-asperiores"
   },
   "taxonomies":[

   ],
   "extras":{
      "title":"Labore reprehenderit dolorem repellendus asperiores.",
      "image":{
         "filename":"kitten.jpg",
         "alt":"A cute kitten",
         "path":"/thumbs/kitten.jpg?w=240&h=160&area=files"
      },
      "excerpt":"Labore reprehenderit dolorem repellendus asperiores. Ut qui dolor debitis esse. Maiores alias …",
      "link":"/homepage/labore-reprehenderit-dolorem-repellendus-asperiores",
      "editLink":"/bolt/edit/1"
   },
   "status":"published",
   "icon":"fa-home",
   "createdAt":"2018-03-25T03:58:54+00:00",
   "modifiedAt":"2018-06-19T16:18:20+00:00",
   "publishedAt":"2019-01-06T19:14:53+00:00",
   "depublishedAt":"2019-02-13T22:01:37+00:00"
}
```

Reponse in beta1 (valid):
```json
{
   "id":1,
   "contentType":"homepage",
   "modifiedAt":"2018-06-19T16:18:20+00:00",
   "publishedAt":"2019-01-06T19:14:53+00:00",
   "fieldValues":{
      "title":"Labore reprehenderit dolorem repellendus asperiores.",
      "image":{
         "filename":"kitten.jpg",
         "alt":"A cute kitten",
         "path":"/thumbs/kitten.jpg?w=240&h=160&area=files"
      },
      "teaser":"Ut qui dolor debitis esse. Maiores alias repellat maxime ut voluptatem. Eum esse reprehenderit expedita necessitatibus. Animi hic eveniet sed et.\n\nEst pariatur incidunt aut iste. Rem aut nihil et ut natus perspiciatis. Dolore veritatis ab sed repellendus quo deleniti qui.\n\nQui laudantium corrupti labore quidem veniam nemo qui. Sit dolore atque et cumque libero. Omnis earum repellendus hic aut vero aut voluptatem. Ea non magni eligendi occaecati numquam.",
      "content":"Minus sed repellendus corporis nemo. Aut aut veniam at aut aliquid. Architecto tempora quia neque numquam voluptas sint est delectus.\n\nUnde voluptatem animi non ut aut dicta. Omnis vero dolorum aliquid laudantium magni asperiores. Et tempora eveniet soluta modi occaecati.\n\nEa dolorum tenetur voluptatum temporibus illo fuga. Quibusdam et doloribus debitis omnis sed. Tempora in aperiam ullam non odit. Praesentium sunt accusantium dolorem commodi labore eum nostrum quia.",
      "contentlink":"Aut ea iusto est iste est odit.",
      "slug":"labore-reprehenderit-dolorem-repellendus-asperiores"
   },
   "taxonomyValues":[

   ],
   "authorName":"Jane Doe",
   "extras":{
      "title":"Labore reprehenderit dolorem repellendus asperiores.",
      "image":{
         "filename":"kitten.jpg",
         "alt":"A cute kitten",
         "path":"/thumbs/kitten.jpg?w=240&h=160&area=files"
      },
      "excerpt":"Labore reprehenderit dolorem repellendus asperiores. Ut qui dolor debitis esse. Maiores alias …",
      "link":"/homepage/labore-reprehenderit-dolorem-repellendus-asperiores",
      "editLink":"/bolt/edit/1"
   }
}
```

**Comment by soyuka** [comment]
Hi, can you give me a way to reproduce this? We have lots of functional tests that would have failed if the results were that different (I hope :p). I'm concerned by the collection here, the response matches an array of arrays?

**Comment by soyuka** [comment]
Btw if you have a custom normalizer, please check priorities!

**Comment by JarJak** [comment]
I don't have any custom normalizer. This is my resource config:
```
/**
 * @ApiResource(
 *     normalizationContext={"groups"={"get_content"}},
 *     collectionOperations={"get"},
 *     itemOperations={"get"}
 * )
 * @ApiFilter(SearchFilter::class)
 * @ORM\Entity(repositoryClass="Bolt\Repository\ContentRepository")
 * @ORM\Table(indexes={
 *     @ORM\Index(name="content_type_idx", columns={"content_type"}),
 *     @ORM\Index(name="status_idx", columns={"status"})
 * })
 * @ORM\HasLifecycleCallbacks
 */
```

You can see how it fails here: https://travis-ci.org/bolt/four/jobs/498919190
And then after conflict added in composer.json, it is green again: https://travis-ci.org/bolt/four/builds/494980649

Changes that made it green again: https://github.com/bolt/four/compare/e5770e48d41b1401c15578611315aedf9d1b1d83...40747f1ebeb9e9c1ac04b4f079178c074e8efdac

You can use Bolt 4 to reproduce the bug, but I'm not sure how to isolate it.

**Comment by dunglas** [comment]
Proposal:

- introfuce a new context option: use_api_platform
- set it to false by default, except for formats fully managed by API Platform (JSON-LD, JSON:API, HAL)
- switch it to true in our SerializerContextBuilder

It will fix such issues.

**Comment by JarJak** [comment]
> introduce a new context option: use_api_platform

so should I just add another serialization group and attach it to same properties and methods as get_content?

> set it to false by default, except for formats fully managed by API Platform (JSON-LD, JSON:API, HAL)
> switch it to true in our SerializerContextBuilder

I'm not sure how to do it, is it documented somewhere?

**Comment by JarJak** [comment]
Here you can see the resource class: https://github.com/bolt/four/blob/master/src/Entity/Content.php

**Comment by soyuka** [comment]
@JarJak the comment from @dunglas is something we need to implement so that you don't face the issue by upgrading. Sorry for the inconvenience.

**Comment by teohhanhui** [comment]
@dunglas:

> introfuce a new context option: use_api_platform

What's the actual issue here?

**Comment by soyuka** [comment]
Actually the issue is that now the ItemNormalizer might be called even though:
  - it's not a resource class (accepts anything since beta-2)
  - it's not one of the following formats: `json-ld`, `json-api`, `hal`

Because the `AbstractItemNormalizer::supportsNormalization` almost always return `true` and that `ApiPlatform\Core\Serializer\ItemNormalizer` has no format check.

**Comment by teohhanhui** [comment]
> the ItemNormalizer might be called even though:
>
>    * it's not a resource class (accepts anything since beta-2)

I'm clearly not caught up on things. But, :scream_cat:

**Comment by JarJak** [comment]
@dunglas @soyuka any news on this one? I literally can't upgrade API Platform to any version above 2.4.0-beta1 :/

**Comment by teohhanhui** [comment]
@JarJak Is the issue still present on the latest `2.4` branch of `api-platform/core`?

If so, please help us add a failing Behat test to our test suite. Thanks!

**Comment by JarJak** [comment]
@teohhanhui I'm sorry but I have no idea how to reproduce the issue in your Behat tests. Can you guide me what have been changed between beta1 and beta2?

**Comment by JarJak** [comment]
This is our configuration if it can help you:
```yaml
api_platform:
    enable_swagger_ui: false # route redirected manually to /bolt/api
    graphql:
        graphiql:
            enabled: true
    mapping:
        paths: ['%kernel.project_dir%/src/Entity']
    formats:
        jsonld:   ['application/ld+json']
        json:     ['application/json']
        jsonapi:  ['application/vnd.api+json']
        html:     ['text/html']
    collection:
            pagination:
                client_items_per_page: true
                items_per_page_parameter_name: 'pageSize'
```

**Comment by JarJak** [comment]
@soyuka @teohhanhui @dunglas I've finally found out what was causing this. See https://github.com/api-platform/core/pull/2732

**Comment by JarJak** [comment]
I can confirm that removing \JsonSerializable from Entity (Resource class) fixes the issue. However there is still something to think about: why in this example a single resource becomes a collection?

**Comment by teohhanhui** [comment]
Should be fixed by api-platform/core#2921 which will be in `2.5`.
